CAPT PBNJ power management integration

pokearu opened the pull request:
## Description
The PR enables automated power on/off of nodes that are made part of the cluster using a CRD driven approach. Currently the BMC information is not part of the Tinkerbell hardware, hence a new CRD is introduced. 

- New CRD `BMC`  and a corresponding controller ` controller/bmc` which is responsible for maintaining the state of the hardware's bmc for the corresponding nodes.
- The bmc controller uses a  `PBNJ client` (similar to the tink client) which enables performing pbnj `PowerRequest` and `DeviceRequest` to manage the nodes.
- The bmc spec has a `SecretReference` field that points to a k8s secret that contains the BMC user creds.
- bmc spec has 2 fields `PowerAction` and `BootDevice` which the controller uses to reconcile the bare metal nodes by performing PBNJ requests. The values for these fields are based from [machine.proto#L40L57](https://github.com/tinkerbell/pbnj/blob/main/api/v1/machine.proto#L40L57).
- During cluster create, when a hardware is selected for a machine, the new `bmcRef` field in the hardware CRD is used to fetch the BMC object and patch the `PowerAction` field and set the desired power state.
- During cluster delete, when a hardware is released, the `bmcRef` field is used to patch the BMC object and set `BootDevice` to PXE and power off the nodes.

## Why is this needed
Tries to address the issues listed on #146.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I have tested these changes by creating and deleting multiple clusters using CAPT with the new modified CRDs.
#### Sample yaml for the new CRDs
```yaml
kind: Hardware
apiVersion: tinkerbell.org/v1alpha1
metadata:
  name: node-1
spec:
  id: 8b8fdc3c-22e8-11ec-9621-0242ac130004
  bmcRef: bmc-node-1
---
kind: BMC
apiVersion: tinkerbell.org/v1alpha1
metadata:
  name: bmc-node-1
spec:
  host: 0.0.0.0
  authSecretRef:
    name: bmc-1-auth
    namespace: test-system
  vendor: Supermicro
---
apiVersion: v1
kind: Secret
metadata:
  name: bmc-1-auth
  namespace: test-system
type: kubernetes.io/basic-auth
data:
  password: QWRtaW4=
  username: YWRtaW4=
```

## How are existing users impacted? What migration steps/scripts do we need?
The existing users are not impacted. The `bmcRef` field in the hardware CRD is optional, and if not specified the power management steps are skipped.

<!--- Fixes a bug, unblocks installation, removes a component of the stack etc -->
<!--- Requires a DB migration script, etc. -->


## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

cprivitere commented:
@pokearu @detiber Seems like there's some issues here, can someone resolve and repush?

pokearu commented:
> @pokearu @detiber Seems like there's some issues here, can someone resolve and repush?

Hi @cprivitere I just rebased and updated some of the files that had issues.

cprivitere commented:
Well, closer at least. Now someone just needs to solve the rest.

pokearu commented:
> Well, closer at least. Now someone just needs to solve the rest.

Updated again! Hopefully its green this time ✅

pokearu commented:
Closing PR as the changes are no longer valid with the Tinkerbell kubernetes model.
Opened new PR #182
